feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-single-selection-base.component.ts

```diff
@@ -0,0 +1,157 @@
+
+  /**
+   * @deprecated since 3.3
+   */
+  constructor();
+
+  constructor(
+    protected quantityService?: ConfiguratorAttributeQuantityService
+  ) {
+    super();
+  }
+
+  /**
+   * Checks if we are supposed to render a quantity control, which
+   * can be derived from the attribute meta data
+   *
+   * @return {boolean} - Display quantity picker?
+   */
+  get withQuantity() {
+    return (
+      this.quantityService?.withQuantity(
+        this.attribute.dataType,
+        this.attribute.uiType
+      ) ?? false
+    );
+  }
+
+  /**
+   * Checks if quantity control should be disabled
+   *
+   * @return {boolean} - Disable quantity picker?
+   */
+  get disableQuantityActions() {
+    return (
+      this.quantityService?.disableQuantityActions(
+        this.attribute.selectedSingleValue
+      ) ?? true
+    );
+  }
+
+  onSelect(value: string): void {
+    this.loading$.next(true);
+
+    const event: ConfigFormUpdateEvent = {
+      changedAttribute: {
+        ...this.attribute,
+        selectedSingleValue: value,
+      },
+      ownerKey: this.ownerKey,
+      updateType: Configurator.UpdateType.ATTRIBUTE,
+    };
+
+    this.selectionChange.emit(event);
+  }
+
+  onDeselect(): void {
+    this.loading$.next(true);
+
+    const event: ConfigFormUpdateEvent = {
+      changedAttribute: {
+        ...this.attribute,
+        selectedSingleValue: '',
+      },
+      ownerKey: this.ownerKey,
+      updateType: Configurator.UpdateType.ATTRIBUTE,
+    };
+
+    this.selectionChange.emit(event);
+  }
+
+  onHandleQuantity(quantity): void {
+    this.loading$.next(true);
+
+    const event: ConfigFormUpdateEvent = {
+      changedAttribute: {
+        ...this.attribute,
+        quantity,
+      },
+      ownerKey: this.ownerKey,
+      updateType: Configurator.UpdateType.ATTRIBUTE_QUANTITY,
+    };
+
+    this.selectionChange.emit(event);
+  }
+
+  onChangeQuantity(eventObject): void {
+    if (!eventObject.quantity) {
+      this.onDeselect();
+    } else {
+      this.onHandleQuantity(eventObject.quantity);
+    }
+  }
+
+  /**
+   *  Extract corresponding quantity parameters
+   *
+   * @param {boolean} disableQuantityActions - Disable quantity actions
+   * @return {ConfiguratorAttributeQuantityComponentOptions} - New quantity options
+   */
+  extractQuantityParameters(): ConfiguratorAttributeQuantityComponentOptions {
+    const quantity: number = this.attribute.quantity ?? 0;
+    return {
+      allowZero: !this.attribute.required,
+      initialQuantity: this.attribute.selectedSingleValue ? quantity : 0,
+      disableQuantityActions$: this.loading$.pipe(
+        map((loading) => {
+          return loading || this.disableQuantityActions;
+        })
+      ),
+    };
+  }
+
+  /**
+   * Extract corresponding price formula parameters
+   *
+   * @return {ConfiguratorPriceComponentOptions} - New price formula
+   */
+  extractPriceFormulaParameters(): ConfiguratorPriceComponentOptions {
```

Comment:
This one is not the shared function. `ConfiguratorAttributeRadioButtonComponent` does not use it. So, it is better to move it to child class.

Reply:
It's true but in the next release `ConfiguratorPriceComponent` will be added to  `ConfiguratorAttributeRadioButtonComponent` as well. In order to avoid breaking changes we decided to move `extractPriceFormulaParameters` method to `ConfiguratorAttributeSingleSelectionBaseComponent` despite the fact that only `ConfiguratorAttributeSingleSelectionBundleComponent` needs it